component/gnap/as/client.go

```diff
@@ -0,0 +1,203 @@
+func NewClient(signer gnap.Signer, httpClient *http.Client, gnapAuthServerURL string) (*Client, error) {
+	if signer == nil {
+		return nil, fmt.Errorf("gnap auth client: missing signer")
+	}
+
+	if httpClient == nil {
+		return nil, fmt.Errorf("gnap auth client: missing http client")
+	}
+
+	if gnapAuthServerURL == "" {
+		return nil, fmt.Errorf("gnap auth client: missing Authorization Server URL")
+	}
+
+	return &Client{
+		signer:            signer,
+		httpClient:        httpClient,
+		gnapAuthServerURL: gnapAuthServerURL,
+	}, nil
+}
+
+// RequestAccess creates a GNAP grant access req then submit it to the server to receive a response with an
+// interact_ref value.
+//nolint:funlen,gocyclo
+func (c *Client) RequestAccess(req *gnap.AuthRequest) (*gnap.AuthResponse, error) {
+	if req == nil {
+		return nil, fmt.Errorf("requestAccess: empty request")
+	}
+
+	mReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: marshal access token error: %w", err)
+	}
+
+	var (
+		sig               []byte
+		marshalledRequest []byte
+	)
+
+	sig, err = c.signer.Sign(mReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: signature error: %w", err)
+	}
+
+	marshalledRequest, err = json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to marshal request: %w", err)
+	}
+
+	requestReader := bytes.NewReader(marshalledRequest)
+
+	url := c.gnapAuthServerURL + gnaprest.AuthRequestPath
+
+	httpReq, err := http.NewRequest(http.MethodPost, url, requestReader)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to build http request: %w", err)
+	}
+
+	httpReq.Header.Add("Content-Type", contentType)
+	// httpReq.Header.Add("Signature-Input", "TODO") // TODO update signature input
+	httpReq.Header.Add("Signature", base64.URLEncoding.EncodeToString(sig))
+
+	r, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to post HTTP request to [%s]: %w", gnaprest.AuthRequestPath, err)
+	}
+
+	defer func() {
+		err = r.Body.Close()
+		if err != nil {
+			logger.Warnf("failed to close http request but it has been processed: %w", err)
+		}
+	}()
+
+	if r.StatusCode >= http.StatusBadRequest {
+		return nil, fmt.Errorf("requestAccess: Auth server replied with invalid Status [%s]: %v",
+			gnaprest.AuthRequestPath, r.Status)
+	}
+
+	respBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: read response failed [%s]: %w", gnaprest.AuthRequestPath, err)
+	}
+
+	gnapResp := &gnap.AuthResponse{}
+
+	err = json.Unmarshal(respBody, gnapResp)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: read response not properly formatted [%s, %w]",
+			gnaprest.AuthRequestPath, err)
+	}
+
+	return gnapResp, nil
+}
+
+// Continue gnap auth request containing interact_ref.
+//nolint:funlen,gocyclo
+func (c *Client) Continue(req *gnap.ContinueRequest) (*gnap.ResponseContinue, error) {
```

Comment:
Why the asymmetry between naming continue request and response models (ContinueRequest vs ResponseContinue) ?

Reply:
ask @Moopli , it was merged in his PR :)

I guess since they are both on the same line here, the asymmetry is apparent

Reply:
Let's add a TODO or something :)

Reply:
Good point, I can fix that in my PR

Reply:
@rolsonquadras wait I see what's going on here, `ResponseContinue` is not the response for `ContinueRequest`. `ResponseContinue` is [this](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-09.html#section-3.1) field inside `AuthResponse`.

@Baha-sk the response to the Continue request is `AuthResponse`.

Reply:
made Continue return `AuthResponse` as well